Support for nested mappings

auipga opened the pull request:
This change adds support for nested mappings.

Example:

``` php
$formMapper
    ->add('car_model', 'choice_field_mask', [
        'choices' => [
            'nice_car' => 'Mercedes',
            'bad_car'  => 'Opel',
        ],
        'map' => [
            'nice_car' => ['nice_car_color'],
            'bad_car' => ['bad_car_color']
        ]
    ])
    ->add('nice_car_color', 'choice_field_mask', [
        'choices' => [
            'silver' => 'silver',
            'black'  => 'black',
        ],
        'map' => [
            'silver' => ['exclusive_color'],
            'black' => [],
        ]
    ])
    ->add('exclusive_color', 'choice', [
        'choices' => [
            'silver1' => 'premium silver',
            'silver2' => 'ultra premium silver',
        ]
    ])
    ->add('bad_car_color', 'choice', [
        'choices' => [
            'blue' => 'blue',
            'red'  => 'red',
        ]
    ])
;
```

`exclusive_color` will only be visible if you select Mercedes AND silver.
`exclusive_color` will be hidden if you change `nice_car_color` (default behaviour) and now also if you change its parent `car_model`.

dbu commented:
hi, thanks for this one?
is this a functionality ported from sonata orm admin? i am a bit reluctant of blowing up javascript in the twig templates. if the orm can do the same, this would much better live in the base SonataAdminBundle than be duplicated in every admin... but for that we probably would have to move a lot of templates to SonataAdminBundle and it would end up being a massive refactoring.
and this thing is useful, so i guess i will merge it if its the same in orm admin.

auipga commented:
hi,
it's not ported. I made this because I couldn't find any behaviour like this but I think it's essential.

dbu commented:
okay. i am trying to get some input from the @sonata-project/contributors to decide if this could be made a general sonata feature instead. if that turns out impracticable, i will merge it in here.

ju1ius commented:
@dbu @auipga There's a PR going on to [move all JS to the admin bundle](https://github.com/sonata-project/SonataAdminBundle/pull/3647), so once it's merged, it should definitely belong there !

ju1ius commented:
@auipga Maybe you could submit a PR against [my branch](https://github.com/ju1ius/SonataAdminBundle/tree/js-refactoring) so we can merge efforts ?

The relevant code is here:
* the [`sonata_type_choice_field_mask_widget`](https://github.com/ju1ius/SonataAdminBundle/blob/js-refactoring/Resources/views/Form/form_admin_fields.html.twig#L464) block
* the [choice-mask](https://github.com/ju1ius/SonataAdminBundle/blob/js-refactoring/Resources/assets/js/sonata/form/type/choice-mask.js) js widget

dbu commented:
thanks for the update @ju1ius . @auipga do you think you can contribute this to the work julius is doing and we close this pull request?

auipga commented:
@dbu in a few weeks yes. at the moment i got not enough time. maybe s.b. else can do this if urgent

core23 commented:
Please use the new PR template @auipga

core23 commented:
Ping @auipga

auipga commented:
If I did understand the PR template and CONTRIBUTING.md correctly, I need to create a new PR against 1.x.
If I try to do so, github wants to add more commits than just mine. Please create the PR by yourself as I don't get it.

Here is my text for the new PR: https://gist.github.com/auipga/c3d604d9a293b5118afa14f5c40c1702
I hope it is about right.

greg0ire commented:
1. Fetch all new commits: `git fetch --all`. ⬇️ 
2. Rebase on what you fetched interactively: ✂️ 
    1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin repository.
    2. A window will show up with many lines, remove every line but the last one, which should correspond to your commit.
   2. If you run into a conflict: 💥 
        1. fix it with `git mergetool`. 😷 
        1. continue on your merry way with `git rebase --continue`. ⏩ 
3. Force push to overwrite all this : `git push --force`. ⬆️

auipga commented:
It has been a bit more difficult (at least for me).
Additional steps: 3, 5, 6
If I skip step 3 and 5 ->rebasing does not show my commits
If I skip step 6 -> when pushing: `fatal: You are not currently on a branch.`

I also changed step 8 (your no. 3) to push to my own remote (because of the write permission).

1. `git clone https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle.git`
2. `cd SonataDoctrinePhpcrAdminBundle`
3. `git remote add auipga https://github.com/auipga/SonataDoctrinePhpcrAdminBundle.git`
4. `git fetch --all`
5. `git checkout auipga/master` ... `You are in 'detached HEAD' state. [...]`
6. `git checkout -b nested HEAD@{0}`
7. `git rebase -i origin/1.x` removed everything but https://github.com/auipga/SonataDoctrinePhpcrAdminBundle/commit/6865ad5f3d40003780c271971f1f8f500b1f2313 and https://github.com/auipga/SonataDoctrinePhpcrAdminBundle/commit/82c0f498b8873a17bb8ea8c0d185d6663cb7c7d0 resulting in `Successfully rebased and updated detached HEAD.`
8. `git push auipga --force`

<!-- 5. `git checkout origin/master` this is where this PR points at, repeated 4 -> lists 78 picks (same count as [here (scroll to the end)](https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/compare/1.x...auipga:master) but not showing my commit -->
